test/mobile/custom_build/build.sh

```diff
@@ -0,0 +1,146 @@
+#!/bin/bash
+###############################################################################
+# This script shows/tests the flow to build libtorch locally with optimized
+# binary size for mobile devices and the flow to integrate it with a simple
+# predictor in c++.
+#
+# There are three custom build types:
+#
+# 1. `TEST_DEFAULT_BUILD=1 ./build.sh` - it is similar to the prebuilt libtorch
+# libraries released for Android and iOS (same CMake build options + host
+# toolchain), which doesn't contain autograd function nor backward ops thus is
+# smaller than full LibTorch.
+#
+# 2. `TEST_CUSTOM_BUILD_STATIC=1 ./build.sh` - it further optimizes libtorch
+# size by only including ops used by a specific model.
+#
+# 3. `TEST_CUSTOM_BUILD_DYNAMIC=1 ./build.sh` - similar as 2) except that it
+# relies on the op dependency graph (instead of static dispatch) to calculate
+# and keep all transitively dependent ops by the model.
+# Note that LLVM_DIR environment variable should be set to the location of
+# LLVM-dev toolchain.
+#
+# Type 2) will be deprecated by type 3) in the future.
+###############################################################################
+
+set -ex -o pipefail
+
+SRC_ROOT="$( cd "$(dirname "$0")"/../../.. ; pwd -P)"
+TEST_SRC_ROOT="${SRC_ROOT}/test/mobile/custom_build"
+BUILD_ROOT="${BUILD_ROOT:-${SRC_ROOT}/build_test_custom_build}"
+mkdir -p "${BUILD_ROOT}"
+cd "${BUILD_ROOT}"
+
+prepare_model_and_dump_root_ops() {
+  cd "${BUILD_ROOT}"
+  MODEL="${BUILD_ROOT}/MobileNetV2.pt"
+  ROOT_OPS="${BUILD_ROOT}/MobileNetV2.yaml"
+
+  if [ ! -f "${MODEL}" ] || [ ! -f "${ROOT_OPS}" ]; then
+    python "${TEST_SRC_ROOT}/prepare_model.py"
+  fi
+}
+
+generate_op_dependency_graph() {
```

Comment:
selected mobile build with dynamic dispatch is not supported in OSS yet so this won't be called now.

When I enable it in the future, it will build libtorch twice - first generate the op dependency graph inline (which will build mobile libtorch into LLVM bitcode), then call the regular mobile build (which will build mobile libtorch again).

I could reuse the output of the other code-analysis CI's output to save some work, but it won't reduce latency of the job. Fortunately it took much less time to build libtorch with mobile CMake options (skipped autograd and caffe2) so it's probably not going to be too bad.

Let me know if you have better suggestions.

Reply:
This is the terminal state of the static analyzer, is that right? Because we cannot use LLVM to generate the op dependency graph without compiling everything, and then we need to build the actual build. Do you have some sense of how long these steps take?

It's kind of irritating that you have to build twice, but assuming you need LLVM bitcode I'm not sure how to get around it. I suppose that for OSS end users, we could directly distribute LLVM bitcode for libtorch, so they don't need to do this step.

Reply:
"oops"